server/proxy/incoming.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/config"
+	"github.com/mattermost/mattermost-plugin-apps/server/mmclient"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+	"github.com/mattermost/mattermost-plugin-apps/utils/httputils"
+)
+
+type Incoming struct {
+	ActingUserID          string
+	PluginID              string
+	ActingUserAccessToken string
+	AdminAccessToken      string
+	SessionID             string
+}
+
+func NewIncomingFromContext(cc apps.Context) Incoming {
+	return Incoming{
+		ActingUserID:          cc.ActingUserID,
+		ActingUserAccessToken: cc.ActingUserAccessToken,
+		AdminAccessToken:      cc.AdminAccessToken,
+	}
+}
+
+func RequireUser(f func(_ http.ResponseWriter, _ *http.Request, in Incoming)) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		actingUserID := req.Header.Get(config.MattermostUserIDHeader)
+		sessionID := req.Header.Get(config.MattermostSessionIDHeader)
+		if actingUserID == "" || sessionID == "" {
+			httputils.WriteError(w, errors.Wrap(utils.ErrUnauthorized, "user ID and session ID are required"))
+			return
+		}
+
+		f(w, req, Incoming{
+			ActingUserID: actingUserID,
+			SessionID:    sessionID,
+		})
+	}
+}
+
+func RequireSysadmin(mm *pluginapi.Client, f func(_ http.ResponseWriter, _ *http.Request, in Incoming)) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		actingUserID := req.Header.Get(config.MattermostUserIDHeader)
+		sessionID := req.Header.Get(config.MattermostSessionIDHeader)
+		if actingUserID == "" || sessionID == "" {
+			httputils.WriteError(w, errors.Wrap(utils.ErrUnauthorized, "user ID and session ID are required"))
+			return
+		}
+		in := Incoming{
+			ActingUserID: actingUserID,
+			SessionID:    sessionID,
+		}
+
+		err := utils.EnsureSysAdmin(mm, actingUserID)
+		if err != nil {
+			httputils.WriteError(w, utils.ErrUnauthorized)
+			return
+		}
+
+		f(w, req, in)
+	}
+}
+
+func RequireSysadminOrPlugin(mm *pluginapi.Client, f func(_ http.ResponseWriter, _ *http.Request, in Incoming)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		pluginID := r.Header.Get(config.MattermostPluginIDHeader)
+		actingUserID := r.Header.Get(config.MattermostUserIDHeader)
+		in := Incoming{
+			PluginID:     pluginID,
+			ActingUserID: actingUserID,
+		}
+
+		if pluginID == "" {
+			sessionID := r.Header.Get(config.MattermostSessionIDHeader)
+			if actingUserID == "" || sessionID == "" {
+				httputils.WriteError(w, errors.Wrap(utils.ErrUnauthorized, "user ID and session ID are required"))
+				return
+			}
+			in.SessionID = sessionID
+
+			err := utils.EnsureSysAdmin(mm, actingUserID)
+			if err != nil {
+				httputils.WriteError(w, utils.ErrUnauthorized)
+				return
+			}
+		}
+
+		f(w, r, in)
+	}
+}
+
+func (p *Proxy) asAdmin(in Incoming) (Incoming, mmclient.Client, error) {
```

Comment:
Is `asAdmin` really the right name here? What does the method return for non-admins?

Reply:
I got rid of this one altogether, we just use the RPC client in the admin methods, since they require a sysadmin anyway for now. The history is that we were missing p.API methods, and resorted to HTTP before, but now there is no need.